File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/DatePickerButton.java
package com.gb.canibuythat.ui;

import android.app.DatePickerDialog;
import android.content.Context;
import android.support.v7.widget.AppCompatButton;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.widget.DatePicker;

import com.gb.canibuythat.util.DateUtils;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

public class DatePickerButton extends AppCompatButton {

    private Calendar originalDate;
    private DatePickerDialog datePickerDialog;
    private GestureDetector detector;
    private DatePickerDialog.OnDateSetListener onDateSetListener = (view, year, monthOfYear, dayOfMonth) ->
            setText(DateUtils.formatDayMonthYear(new GregorianCalendar(year, monthOfYear, dayOfMonth).getTime()));

    public DatePickerButton(Context context) {
        super(context);
        init();
    }

    public DatePickerButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public DatePickerButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        detector = new GestureDetector(getContext(), new GestureDetector.SimpleOnGestureListener() {
            @Override
            public boolean onSingleTapConfirmed(MotionEvent e) {
                getDatePickerDialog().show();
                return false;
            }
        });
        originalDate = Calendar.getInstance();
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent e) {
        detector.onTouchEvent(e);
        return super.dispatchTouchEvent(e);
    }

    private DatePickerDialog getDatePickerDialog() {
        if (datePickerDialog == null) {
            datePickerDialog = DateUtils.getDatePickerDialog(getContext(), onDateSetListener, originalDate);
        }
        return datePickerDialog;
    }

    public void setDate(Date date) {
        originalDate.setTime(date);
        if (datePickerDialog != null) {
            getDatePickerDialog().updateDate(originalDate.get(Calendar.YEAR),
                    originalDate.get(Calendar.MONTH),
                    originalDate.get(Calendar.DAY_OF_MONTH));
        }
    }

    public Date getSelectedDate() {
        if (datePickerDialog != null) {
            DatePicker datePicker = datePickerDialog.getDatePicker();
            return new GregorianCalendar(datePicker.getYear(), datePicker.getMonth(), datePicker.getDayOfMonth()).getTime();
        }
        return originalDate.getTime();
    }
}

File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/DatePickerButton.kt
package com.gb.canibuythat.ui

import android.app.DatePickerDialog
import android.content.Context
import android.support.v7.widget.AppCompatButton
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.MotionEvent
import com.gb.canibuythat.util.createDatePickerDialog
import com.gb.canibuythat.util.formatDayMonthYearWithPrefix
import java.time.LocalDate

class DatePickerButton : AppCompatButton {

    private lateinit var originalDate: LocalDate
    private var datePickerDialog: DatePickerDialog? = null
    private lateinit var detector: GestureDetector

    val selectedDate: LocalDate
        get() {
            if (datePickerDialog != null) {
                val datePicker = datePickerDialog!!.datePicker
                return LocalDate.of(datePicker.year, datePicker.month + 1, datePicker.dayOfMonth)
            }
            return originalDate
        }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        detector = GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
            override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
                getDatePickerDialog().show()
                return false
            }
        })
        originalDate = LocalDate.now()
    }

    override fun dispatchTouchEvent(e: MotionEvent): Boolean {
        detector.onTouchEvent(e)
        return super.dispatchTouchEvent(e)
    }

    private fun getDatePickerDialog(): DatePickerDialog {
        if (datePickerDialog == null) {
            datePickerDialog = createDatePickerDialog(context, originalDate) { _, date ->
                text = date.formatDayMonthYearWithPrefix()
            }
        }
        return datePickerDialog!!
    }

    fun setDate(date: LocalDate) {
        originalDate = date
        if (datePickerDialog != null) {
            getDatePickerDialog().updateDate(originalDate.year, originalDate.monthValue, originalDate.dayOfMonth)
        }
    }
}
